src/callbacks/callbacks.jl

```diff
@@ -333,7 +333,13 @@ NVTX.@annotate function save_state_to_disk_func(integrator, output_dir)
     output_file = joinpath(output_dir, "day$day.$sec.hdf5")
     comms_ctx = ClimaComms.context(integrator.u.c)
     hdfwriter = InputOutput.HDF5Writer(output_file, comms_ctx)
-    InputOutput.HDF5.write_attribute(hdfwriter.file, "time", t) # TODO: a better way to write metadata
+    # TODO: a better way to write metadata
+    InputOutput.HDF5.write_attribute(hdfwriter.file, "time", t)
+    InputOutput.HDF5.write_attribute(
+        hdfwriter.file,
+        "atmos_model_hash",
+        hash(p.atmos),
```

Comment:
```suggestion
        hash(typeof(p.atmos)),
```
I think we might want to just make sure that the types are the same? Otherwise this might be too brittle (e.g., if `p.atmos` depends on `t`/`t_0`, is `t` the restart time?)

Reply:
`p.atmos` is the `AtmosModel`, so, ideally, it should pretty much only contain types used for dispatch:
```
    model_config::MC = nothing
    moisture_model::MM = nothing
    precip_model::PM = nothing
    cloud_model::CM = nothing
    call_cloud_diagnostics_per_stage::CCDPS = nothing
    forcing_type::F = nothing
    subsidence::S = nothing
    radiation_mode::RM = nothing
    ls_adv::LA = nothing
    external_forcing::EXTFORCING = nothing
    edmf_coriolis::EC = nothing
    advection_test::AT = nothing
    tendency_model::TM = nothing
    edmfx_model::EDMFX = nothing
    turbconv_model::TCM = nothing
    non_orographic_gravity_wave::NOGW = nothing
    orographic_gravity_wave::OGW = nothing
    hyperdiff::HD = nothing
    vert_diff::VD = nothing
    diff_mode::DM = nothing
    sgs_adv_mode::SAM = nothing
    viscous_sponge::VS = nothing
    rayleigh_sponge::RS = nothing
    sfc_temperature::ST = nothing
    insolation::IN = nothing
    surface_model::SM = nothing
    surface_albedo::SA = nothing
    numerics::NUM = nothing
```

If these types contain extra information, I think we should check for consistency of that too.

Reply:
> `p.atmos` is the `AtmosModel`, so, ideally, it should pretty much only contain types used for dispatch:

But it's not only used for dispatch. For example, a Rayleigh sponge is a property of the atmos model, which is computed like this:

```julia
function get_rayleigh_sponge_model(parsed_args, params, ::Type{FT}) where {FT}
    rs_name = parsed_args["rayleigh_sponge"]
    return if rs_name in ("false", false)
        nothing
    elseif rs_name in ("true", true, "RayleighSponge")
        zd = params.zd_rayleigh
        α_uₕ = params.alpha_rayleigh_uh
        α_w = params.alpha_rayleigh_w
        RayleighSponge{FT}(; zd, α_uₕ, α_w)
    else
        error("Uncaught rayleigh sponge model `$rs_name`.")
    end
end
```
That means that users cannot restart with a stronger sponge (perhaps to damp late instabilities). This is just one example, but I believe that there are more submodels to the atmos model that have values.

`typeof(p.atmos)` is arguably to strict still. For example, it prohibits restarting a simulation that started with Float32 and changes to Float64. We could further transform, or write some sort of custom hash, but generally I think we'll want something a bit more flexible.

> If these types contain extra information, I think we should check for consistency of that too.

Check for what sort of consistency?

Reply:
> That means that users cannot restart with a stronger sponge (perhaps to damp late instabilities). This is just one example, but I believe that there are more submodels to the atmos model that have values.

That's correct. At this point, the safest thing to do is to disallow this behavior because we have no guarantee that the content of `p.atmos` would not lead to configurations that are internally inconsistent. 

I agree we want more flexibility, but I think that should be done as a second step and more carefully: first, we allow checkpoints with restrictions, and then we can work on removing those restrictions.

> Check for what sort of consistency?

`Y` and `p` have to be valid `Y` and `p` for the given `AtmosModel`. A case where they would not be valid is if you changed gravitational acceleration.

Reply:
Physics component mismatches will be caught elsewhere upon generating the `simulation` object, e.g. `prognostic_tke` or `0M / 1M`, `equil / nonequil` (each of these options requiring different variables in the state `Y`); I agree that displaying mismatched elements in these error messages would be helpful! It seems reasonable to me to restrict this first pass to focus on "history critical" elements, i.e. restrict the possibility of updating tendencies / model parameters upon restart. Do `checkpoint-restart` and more flexible `data-as-init` use-cases need to be entirely separate (e.g. higher-res interpolations or timestepper swaps)?

Reply:
Ok, I'm fine with this so long it also doesn't prevent someone from manually restarting without this restriction, because that would be removing existing functionality--crude as it may be.

Reply:
I can turn it into a warning